patches/chromium/fix_harden_blink_scriptstate_maybefrom.patch

```diff
@@ -0,0 +1,104 @@
+leading to the following CHECK failure.
+
+```
+script_state.h(169)] Security Check Failed: script_state
+```
+
+This patch adds a hook to allow embedders confirm if a given context is
+something created by them.
+
+PS: We are not performing a check like
+
+```
+ScriptState* script_state =
+    static_cast<ScriptState*>(context->GetAlignedPointerFromEmbedderData(
+        kV8ContextPerContextDataIndex));
+if (!script_state) {
+  return nullptr;
+}
+```
+
+since in 32-bit builds which does not have v8 sandbox enabled unlike 64-bit builds,
+the embedder data slot will not lazy initialize indexes in the former. This means
+accessing uninitialized lower indexes can return garbage values that cannot be null checked.
+Refer to v8::EmbedderDataSlot::store_aligned_pointer for context.
+
+diff --git a/third_party/blink/renderer/platform/bindings/script_state.cc b/third_party/blink/renderer/platform/bindings/script_state.cc
+index 7ff8785cd64c1264a88f91f7bd3292c6943f58ea..2f60dcf26b175727f447edfc94236c1c3f7e7855 100644
+--- a/third_party/blink/renderer/platform/bindings/script_state.cc
++++ b/third_party/blink/renderer/platform/bindings/script_state.cc
+@@ -12,6 +12,7 @@
+ namespace blink {
+ 
+ ScriptState::CreateCallback ScriptState::s_create_callback_ = nullptr;
++ScriptState::ContextEmbedderDataCallback ScriptState::s_context_embedder_data_callback_ = nullptr;
+ 
+ // static
+ void ScriptState::SetCreateCallback(CreateCallback create_callback) {
+@@ -20,6 +21,14 @@ void ScriptState::SetCreateCallback(CreateCallback create_callback) {
+   s_create_callback_ = create_callback;
+ }
+ 
++// static
++void ScriptState::SetContextEmbedderDataCallback(
++    ContextEmbedderDataCallback callback) {
++  DCHECK(callback);
++  DCHECK(!s_context_embedder_data_callback_);
++  s_context_embedder_data_callback_ = callback;
++}
++
+ // static
+ ScriptState* ScriptState::Create(v8::Local<v8::Context> context,
+                                  scoped_refptr<DOMWrapperWorld> world,
+diff --git a/third_party/blink/renderer/platform/bindings/script_state.h b/third_party/blink/renderer/platform/bindings/script_state.h
+index 7109852950cde0a6553000421faacefb39366b41..16957f0ab51a36fde2112ac97cf9352e752af301 100644
+--- a/third_party/blink/renderer/platform/bindings/script_state.h
++++ b/third_party/blink/renderer/platform/bindings/script_state.h
+@@ -123,6 +123,8 @@ class PLATFORM_EXPORT ScriptState : public GarbageCollected<ScriptState> {
+     v8::Local<v8::Context> context_;
+   };
+ 
++  using ContextEmbedderDataCallback = bool (*)(v8::Local<v8::Context>);
++
+   static ScriptState* Create(v8::Local<v8::Context>,
+                              scoped_refptr<DOMWrapperWorld>,
+                              ExecutionContext*);
+@@ -181,9 +183,15 @@ class PLATFORM_EXPORT ScriptState : public GarbageCollected<ScriptState> {
+         kV8ContextPerContextDataIndex) {
+       return nullptr;
+     }
++    if (s_context_embedder_data_callback_ &&
```

Comment:
The comment of `ScriptState::MaybeFrom` says:

```c++
  // For use when it is not absolutely certain that the v8::Context is
  // associated with a ScriptState. This is necessary in unit tests when a
  // v8::Context is created directly on the v8 API without going through the
  // usual blink codepaths.
```

I think what we need to do is to reliably check whether a context comes from blink, instead of checking if it is a node vm context, since the latter check may be broken when using a native module that creates v8 context.

An alternative solution is to add a tag to the context in `blink::V8PerContextData` (similar to what Node does with `kNodeContextTag`), and then check the tag in `ScriptState::MaybeFrom`. In this way we can upstream this patch to blink, and get rid of the patch in Node. This will also help fix other crashes caused by vm context like the one https://github.com/electron/electron/pull/38912 is trying to fix.

Reply:
> since the latter check may be broken when using a native module that creates v8 context.

Ah didn't think about this caller. I will rework the change. Thanks for checking!